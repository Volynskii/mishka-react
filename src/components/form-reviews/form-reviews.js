import React, {useEffect, useState} from "react";
import './form-reviews.scss';
import {useDispatch} from "react-redux";
import {createReview} from "../../store/comments/actions";
import {delay} from "../../helpers/delay";

export const FormReviews = ({isOpened, closePopup,openSuccessPopup}) => {
    const dispatch = useDispatch();
    const [name,setName] = useState('');
    const [email, setEmail] = useState('');
    const [comment, setComment] = useState('');

    const onSubmit =  (evt) => {
        evt.preventDefault();
        evt.currentTarget.reset();
            const review = { name, email, comment };
            dispatch(createReview(review));
            closePopup();
            openSuccessPopup();
    };

    useEffect(() => {
        const onEscKeyDown = (evt) => {
            if (evt.key === `Escape`) {
                   closePopup();
            }
        };
        document.addEventListener(`keydown`, onEscKeyDown);
        return () => document.removeEventListener(`keydown`, onEscKeyDown);
    }, []);

    const onConfirmClick = (evt) => {
        evt.stopPropagation();
        if (evt.target === evt.currentTarget) {
            closePopup();
        }
    };
    return (
        <>
            <div style={{display: isOpened}} onClick={onConfirmClick}  className="popup">
        <form className="form-popup" onSubmit={onSubmit} >
        <section className="form-reviews__personal-data">
            <div className="form-reviews__personal-data__wrapper">
                <ul className="form-reviews__personal-data__list">

                    <li className="form-reviews__personal-data__list__item">
                        <label className="label-name" htmlFor="name">Имя: </label>
                        <input
                            onChange={(evt) =>
                                evt.target.value.includes('@') ? setName(evt.target.value) : 'пусто'
                                // setName(evt.target.value.includes('@'))
                            }
                            type="text" id="name" placeholder="Введите ваше имя*" />
                    </li>

                    <li className="form-reviews__personal-data__list__item">
                        <label className="label-family-name" htmlFor="family-name">email: </label>
                        <input
                            onChange={(evt) => setEmail(evt.target.value)}
                            type="text" id="family-name" placeholder="Укажите фамилию*"/>
                    </li>

                        <textarea
                            onChange={(evt) => setComment(evt.target.value)}
                            className="form-reviews__textarea" id="patronymic" placeholder="Пожалуйста, оставьте отзыв!"/>
                    <div className="form-reviews__button-container">
                        <input type="submit" className="form-reviews__button-container__button" value="Оставить отзыв"/>
                        <strong className="form-reviews__button-container__span">*Поля обязательны для заполнения</strong>
                    </div>

                </ul>
            </div>
        </section>
        </form>

            </div>
            </>
    )
};
