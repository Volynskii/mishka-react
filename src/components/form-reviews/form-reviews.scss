@import "../../utilities/variables";

@media (min-width: $desktop-width-min) {

  .form-popup {
    z-index: 200;
    position: absolute;
  }

  .popup {
    position: fixed;
    z-index: 100;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    overflow: hidden;
    background-color: rgba(0, 0, 0, 0.4);
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 14px;
  }


  .popup__container {
    display: flex;
    flex-direction: column;
    width: 1000px;
    min-height: 550px;
    padding: 20px;
    border-radius: 5px;
    background-color: #f8f8f8;
    justify-content: space-between;
    border: 1px solid #c1c1c1;
    box-shadow: 0 3px 6px rgba(0, 0, 0, 0.2);
  }
  .form-reviews__personal-data {
    background-color: #ffffff;
    display: flex;
    justify-content: flex-start;
    align-items: flex-start;
    width: 1050px;
    min-height: 370px;
    border: 1px solid;
    margin-left: 20px;
    padding-top: 20px;
    padding-bottom: 20px;

    font-size: 19px;
    font-weight: bold;
  }
  .form-reviews__personal-data__wrapper {
    flex-grow: 1;
    display: flex;
    min-height: 210px;
  }
  .form-reviews__personal-data__list {
    width: 700px;
    margin: 0;
    margin-left: 85px;
    padding: 0;
    list-style-type: none;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
  }
  .form-reviews__personal-data__list__item {
    font-size: 19px;
    width: 100%;
    height: 57px;
    display: flex;
    justify-content: flex-start;
    align-items: center;
  }

  .label-name,
  .label-family-name,
  .label-patronymic {
    margin: 0;
    margin-right: 26px;
  }
  input#name,
  input#family-name {
    padding-left: 18px;
    transition: $transition;
    margin: 0;
    flex-grow: 1;
    height: 52px;
    border: 0;
    border-bottom: 2px solid #555555;
  }

  input#name:focus,
  input#family-name:focus {
    flex-grow: 1;
    height: 52px;
    border: 0;
    outline: 0;
    border-bottom: 2px solid #96e0d1;
  }

  .form-reviews__textarea {
margin-top: 50px;
    height: 100px;
  }
  .form-reviews__textarea:placeholder-shown {
text-align: center;
    font-size: 24px;
  }


  .form-reviews__button-container {
    margin-top: 50px;
    width: 700px;
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
  .form-reviews__button-container__button {
    transition: $transition;
    background-color: transparent;
    height: 52px;
    font-size: 17px;
    line-height: 18px;
    font-weight: bold;
    border: 2px solid #444444;
  }
  .form-reviews__button-container__button:hover {
    background-color: #444444;
    color: #ffffff;
  }
  .form-reviews__button-container__button:active {
    opacity: 0.5;
  }
  .form-reviews__button-container__span {
    font-size: 17px;
    font-weight: normal;
  }
};
