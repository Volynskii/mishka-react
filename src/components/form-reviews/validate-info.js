export default function validateInfo(values) {
    let errors = {};

    //Name
    if (!values.userName.trim()) {
        errors.userName = "Username required"
    } else if (values.userName.trim().length < 1) {
        errors.userName = "Username is too short!"
    }
    //Email
    if (!values.email) {
        errors.email = "Email required"
    } else if (!/\S+@\S+\.\S+/.test(values.email)) {
        errors.email = 'Email address is invalid';
    }

    //Comment
    if(!values.comment) {
        errors.comment = "Comment is required"
    } else if (values.comment.length < 6) {
        errors.comment = 'Commend need to be 6 characters of more'
    }

    return errors;
}
