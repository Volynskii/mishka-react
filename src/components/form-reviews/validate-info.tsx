type formValuesProps = {
    values: {
    userName: string,
    email: string,
    comment: string
}

};

export const validateInfo: React.FC<formValuesProps> = ({values}) => {
    let errors:any = {};
    var illegalChars = /\W/; // allow letters, numbers, and underscores
    //Name
    if (!values.userName.trim()) {
        errors.userName = "Username required"
    }   else if ((values.userName.length < 5) || (values.userName.length > 15)) {
        errors.userName = "&bull; Username must have 5-15 characters";
    } else if (illegalChars.test(values.userName)) {
        errors.userName = "&bull; Please enter valid Username. Use only numbers and alphabets";
    }
    //Email
    if (!values.email) {
        errors.email = "Email required"
    } else if (!/\S+@\S+\.\S+/.test(values.email)) {
        errors.email = 'Email address is invalid';
    }

    //Comment
    if(!values.comment) {
        errors.comment = "Comment is required"
    } else if (values.comment.trim().length < 6) {
        errors.comment = 'Commend need to be 6 characters of more'
    } else if (typeof values.comment === 'number') {
        console.log('COMMENT TYPE', typeof values.comment)
        errors.comment = 'Comment is supposed to have not only numbers'
    }

    return errors;
};
 export default validateInfo;
