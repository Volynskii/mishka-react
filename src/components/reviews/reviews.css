.reviews {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 1050px;
    margin: 0;
    margin-left: auto;
    margin-right: auto;
    margin-bottom: 67px;
    height: 457px;
    border: 1px solid;
    background-image: url("./images/review.png");
    background-repeat: no-repeat;
    background-position: left;
}
.reviews__wrapper {
    height: 275px;
    width: 876px;
    position: relative;
}
.reviews__headline {
    position: relative;
    bottom: 8px;
    font-size: 53px;
    line-height: 53px;
    font-weight: bold;
    margin: 0;
    padding: 0;
    text-align: right;
}
.reviews__button {
    color:#444444;
    font-size: 17px;
    line-height: 20px;
    font-weight: bold;
    background-color:transparent;
    width: 170px;
    height: 48px;
    margin: 0;
    margin-left: auto;
    padding: 0;
    position: absolute;
    top: 0;
    right: 0;
}
.reviews__comment {
    font-size: 17px;
    line-height: 30px;
    margin: 0;
    margin-top: 35px;
}
.reviews__author {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    height: 45px;
    position: absolute;
    left: 0;
    bottom: 10px;
    list-style-type: none;
    margin: 0;
    padding: 0;
}
.reviews__author__name {
    font-size: 17px;
    font-weight: bold;
}
.reviews__slider {
    width: 137px;
    height: 45px;
    display: flex;
    justify-content: space-between;
    position: absolute;
    right: 0;
    bottom: 0;
    list-style-type: none;
    margin: 0;
    padding: 0;
}
.reviews__slider__left {
width: 50px;
    background-image: url("./images/huge-arrow-left.png");
    background-position: top;
    background-repeat: no-repeat;
}
.reviews__slider__right {
    width: 50px;
    background-image: url("./images/huge-arrow-right.png");
    background-position: top;
    background-repeat: no-repeat;
}

