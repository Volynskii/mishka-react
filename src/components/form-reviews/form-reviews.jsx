import React, {useEffect} from "react";
import './form-reviews.scss';
import {useDispatch} from "react-redux";
import {createReview} from "../../store/comments/actions";
import useForm from "../../helpers/useForm";
import validate from "./validate-info"


export const FormReviews = ({isOpened, closePopup,openSuccessPopup}) => {

    const { handleChange, values, handleSubmit, errors} = useForm(validate);

    const dispatch = useDispatch();

    const onSubmit =  (evt) => {

        // check if errors exist
         handleSubmit(evt);
        // check if values exist and inputs are fielded
        const errorsAmount = Object.keys(errors).length;
        const eachInputHasValues = Object.keys(values).every((k) => values[k]);

        if (!errorsAmount && eachInputHasValues) {
            let review = values;
            dispatch(createReview(review));
            evt.currentTarget.reset();
            closePopup();
            openSuccessPopup();
        }
    };


    useEffect(() => {
        const onEscKeyDown = (evt) => {
            if (evt.key === `Escape`) {
                closePopup();
            }
        };
        document.addEventListener(`keydown`, onEscKeyDown);
        return () => document.removeEventListener(`keydown`, onEscKeyDown);
    }, [closePopup]);

    const onConfirmClick = (evt) => {
        evt.stopPropagation();
        if (evt.target === evt.currentTarget) {
            closePopup();
        }
    };
    return (
        <>
            <div style={{display: isOpened}} onClick={onConfirmClick} className="popup">
                <form className="form-popup"
                      onSubmit={(evt) => onSubmit(evt)}>
                    <section className="form-reviews__personal-data">
                        <div className="form-reviews__personal-data__wrapper">
                            <ul className="form-reviews__personal-data__list">

                                <li className="form-reviews__personal-data__list__item">
                                    <label className="label-name" htmlFor="name">Имя: </label>
                                    <input
                                        name="userName"
                                        onChange={handleChange}
                                        type="text" id="name" placeholder="Введите ваше имя*"/>
                                </li>
                                {errors.userName && <p className="error-message">{errors.userName}</p>}
                                <li className="form-reviews__personal-data__list__item">
                                    <label className="label-family-name" htmlFor="email">email: </label>
                                    <input
                                        name="email"
                                        onChange={handleChange}
                                        type="text" id="email" placeholder="Укажите фамилию*"/>
                                </li>
                                {errors.email && <p className="error-message">{errors.email}</p>}
                                <textarea
                                    name="comment"
                                    onChange={handleChange}
                                    className="form-reviews__textarea" id="patronymic"
                                    placeholder="Пожалуйста, оставьте отзыв!"/>
                                {errors.comment && <p className="error-message">{errors.comment}</p>}
                                <div className="form-reviews__button-container">
                                    <input type="submit" className="form-reviews__button-container__button"
                                           value="Оставить отзыв"/>
                                    <strong className="form-reviews__button-container__span">*Поля обязательны для
                                        заполнения</strong>
                                </div>

                            </ul>
                        </div>
                    </section>
                </form>

            </div>
        </>
    );
};
