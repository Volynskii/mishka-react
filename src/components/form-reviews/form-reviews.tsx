import React, {FormEvent, useEffect, useState} from "react";
import './form-reviews.scss';
import {useDispatch} from "react-redux";
import {createReview} from "../../store/comments/actions";
import useForm from "../../helpers/useForm";
import validate from "./validate-info"

type reviewsContainerProps = {
    isOpened: string;
    closePopup: () => void;
    openSuccessPopup: () => void;
}
export const FormReviews: React.FC<reviewsContainerProps> = ({isOpened, closePopup,openSuccessPopup}) => {

    const { handleChange, values, handleSubmit, errors }: {
        handleChange: () => void;
        values: {
            userName: string;
            email: string;
            comment: string;
        }
        handleSubmit: () => void;
        errors: any


    } = useForm(validate);

    const dispatch = useDispatch();

    const onSubmit =  (evt: any) => {
        handleSubmit();
        const errorsAmount = Object.keys(errors).length;
        // @ts-ignore
        const eachInputHasValues = Object.keys(values).every((k): Array<string> => values[k]);
        if (!errorsAmount && eachInputHasValues) {
            let review = values;
            dispatch(createReview(review));
            evt.currentTarget.reset();
            closePopup();
            openSuccessPopup();
        }
    };


    useEffect(() => {
        const onEscKeyDown = (evt: KeyboardEvent) => {
            if (evt.key === `Escape`) {
                closePopup();
            }
        };
        document.addEventListener(`keydown`, onEscKeyDown);
        return () => document.removeEventListener(`keydown`, onEscKeyDown);
    }, [closePopup]);

    const onConfirmClick:(evt: any) => void = (evt: any) => {
        evt.stopPropagation();
        if (evt.target === evt.currentTarget) {
            closePopup();
        }
    };
    return (
        <>
            <div style={{display: isOpened}} onClick={onConfirmClick}  className="popup">
                <form className="form-popup"
                      onSubmit={onSubmit} >
                    <section className="form-reviews__personal-data">
                        <div className="form-reviews__personal-data__wrapper">
                            <ul className="form-reviews__personal-data__list">

                                <li className="form-reviews__personal-data__list__item">
                                    <label className="label-name" htmlFor="name">Имя: </label>
                                    <input
                                        name="userName"
                                        onChange={handleChange}
                                        type="text" id="name" placeholder="Введите ваше имя*" />
                                </li>
                                {errors.userName && <p className="error-message">{errors.userName}</p>}
                                <li className="form-reviews__personal-data__list__item">
                                    <label className="label-family-name" htmlFor="family-name">email: </label>
                                    <input
                                        name="email"
                                        onChange={handleChange}
                                        type="text" id="family-name" placeholder="Укажите фамилию*"/>
                                </li>
                                {errors.email && <p className="error-message">{errors.email}</p>}
                                <textarea
                                    name="comment"
                                    onChange={handleChange}
                                    className="form-reviews__textarea" id="patronymic" placeholder="Пожалуйста, оставьте отзыв!"/>
                                {errors.comment && <p className="error-message">{errors.comment}</p>}
                                <div className="form-reviews__button-container">
                                    <input type="submit" className="form-reviews__button-container__button" value="Оставить отзыв"/>
                                    <strong className="form-reviews__button-container__span">*Поля обязательны для заполнения</strong>
                                </div>

                            </ul>
                        </div>
                    </section>
                </form>

            </div>
        </>
    )
};
