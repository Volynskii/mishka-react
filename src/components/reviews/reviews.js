import './reviews.scss'
import {connect} from "react-redux"
import React, {Component} from "react";


 class Reviews extends Component {
    // state = {
    //     index:0,
    // };

render() {

    console.log( 'APP',this.props)
    const {comments,index,buttonLeftIsDisabled,buttonRightIsDisabled} = this.props;
    return (
        <article className="reviews">
            <div className="reviews__wrapper">
                <span className="reviews__headline">Отзывы</span>
                <button className="reviews__button">Написать</button>

                <p className="reviews__comment">Я молодая мама, но всё равно фотографии не набирали
                    и близко такого количества лайков,
                    как у популярных инстамамочек. В отчаянии,
                    я накупила аксессуаров и игрушек в Мишке,
                    и мои фотографии сразу стали более стильными,
                    а также набирают больше лайков!</p>

                <ul className="reviews__author">
                    <li  className="reviews__author__name">{comments[index].author}</li>
                    <li className="reviews__author__email">@misssssiskras</li>
                </ul>
                <section className="reviews__slider">
                    {/*<ButtonNext   commentsLength={comments.length} />*/}
                    {/*<ButtonPrev  commentsLength={comments.length}/>*/}
                    <button disabled={buttonLeftIsDisabled}   onClick={() => this.props.onNext()} className="reviews__slider__left"/>
                    <button  onClick={() => this.props.onSub()}  className="reviews__slider__right"/>
                </section>


            </div>
        </article>
    )
}
};

 function mapStateToProps(state,) {
     return {
         index: state.index,
         buttonLeftIsDisabled:state.buttonLeftIsDisabled,
         buttonRightIsDisabled:state.buttonRightIsDisabled,

     }
 }
 function mapDispatchToProps(dispatch) {
return {
    onNext: () => dispatch({type: 'NEXT'}),
    onSub: () => dispatch({type: 'PREV'})
}
 }
 export default connect(mapStateToProps,mapDispatchToProps) (Reviews);
