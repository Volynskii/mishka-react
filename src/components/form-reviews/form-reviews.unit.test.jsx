import React from "react";
import {mount, shallow} from "enzyme";
import {Provider} from "react-redux";
import {useReviews} from "../../modules/reviews/useReviews";
import configureStore from "redux-mock-store";
import {act} from "react-dom/test-utils";
import renderer from "react-test-renderer";
import {FormReviews} from "./form-reviews";
import {Reviews} from "../../modules/reviews/reviews/reviews";
import * as redux from "react-redux";

let comments = [
    {
        comment: 'Я молодая мама,' +
            ' но всё равно фотографии' +
            ' не набирали и близко такого ' +
            'количества лайков, как у ' +
            'популярных инстамамочек.' +
            ' В отчаянии, я накупила ' +
            'аксессуаров и игрушек в ' +
            'Мишке, и мои фотографии ' +
            'сразу стали более стильными,' +
            ' а также набирают больше лайков!',
        author: 'Анастасия Красильникова',
        email: '@misssssiskras'
    },
    {
        comment: 'Я молодой парень и мои мои' +
            ' фотографии обычно не набирали' +
            ' большого количества лайков, ' +
            'но как только я купил игрушки ' +
            'разных форматов цветов и форматов' +
            ' количество лайков пошло в гору!' +
            ' Я очень доволен! ',
        author: 'Михаил Соколов!',
        email: 'sokolov@gmail.com'
    },
    {
        comment: 'Я очень доволен данными' +
            ' товарами,товары имеют' +
            ' чрезвычайно высокое качество,' +
            ' ' +
            'заказал себе, друзьям, бабушке,дедушке, папе и маме!',

        author: 'Илья Перевалов!',
        email: 'perevalov@gmail.com'
    }

];

const initialState = {
    comments: comments,
    index:0,

};
const mockStore = configureStore();
const store = mockStore(initialState);

let results;
const renderHook = (hook) => {

    function HookWrapper() {
        results = hook();
        return null;
    }
    mount(
        <Provider store={store}>
            <HookWrapper/>
        </Provider>
    );
    return results;
};

renderHook(useReviews)

const initialProps = {
    isOpened:results.isOpened,
    closePopup: results.onClosePopup,
    openSuccessPopup: results.onOpenSuccessPopup,

};

const container = mount(
    <Provider store={store}>
        <FormReviews {...initialProps}/>
    </Provider>
);

const useDispatchSpy = jest.spyOn(redux, 'useDispatch');
const mockDispatchFn = jest.fn()
useDispatchSpy.mockReturnValue(mockDispatchFn);
describe('Reviews work', () => {

    it('should render Reviews Component', ()=> {
        const submitButton = container.find('#name');

        const dom = <FormReviews isOpened={results.display} closePopup={results.onClosePopup}
                                 openSuccessPopup={results.openSuccessPopup} />;
        const component = renderer.create(dom);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        // const dom = <FormReviews {...initialProps} />;
        // const component = renderer.create(dom);
        // const tree = component.toJSON();
        // expect(tree).toMatchSnapshot();
    });

});
