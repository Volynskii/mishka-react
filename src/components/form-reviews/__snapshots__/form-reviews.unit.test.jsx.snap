// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reviews work should render Reviews Component 1`] = `
<div
  className="popup"
  onClick={[Function]}
  style={
    Object {
      "display": "none",
    }
  }
>
  <form
    className="form-popup"
    onSubmit={[Function]}
  >
    <section
      className="form-reviews__personal-data"
    >
      <div
        className="form-reviews__personal-data__wrapper"
      >
        <ul
          className="form-reviews__personal-data__list"
        >
          <li
            className="form-reviews__personal-data__list__item"
          >
            <label
              className="label-name"
              htmlFor="name"
            >
              Имя: 
            </label>
            <input
              id="name"
              name="userName"
              onChange={[Function]}
              placeholder="Введите ваше имя*"
              type="text"
            />
          </li>
          <li
            className="form-reviews__personal-data__list__item"
          >
            <label
              className="label-family-name"
              htmlFor="family-name"
            >
              email: 
            </label>
            <input
              id="family-name"
              name="email"
              onChange={[Function]}
              placeholder="Укажите фамилию*"
              type="text"
            />
          </li>
          <textarea
            className="form-reviews__textarea"
            id="patronymic"
            name="comment"
            onChange={[Function]}
            placeholder="Пожалуйста, оставьте отзыв!"
          />
          <div
            className="form-reviews__button-container"
          >
            <input
              className="form-reviews__button-container__button"
              type="submit"
              value="Оставить отзыв"
            />
            <strong
              className="form-reviews__button-container__span"
            >
              *Поля обязательны для заполнения
            </strong>
          </div>
        </ul>
      </div>
    </section>
  </form>
</div>
`;
